### PR TITLE
Register scripts during *_enqueue_scripts hook instead

### LINKED ISSUE #361: wp_register_script causes a PHP Notice
I have no idea why I just started getting this, but I just updated this [VIP Go repo](https://github.com/Automattic/vip-go-mu-plugins) in my local environment and I suddenly started getting two PHP notices. Turns out the notice came from these functions:

https://github.com/WordPress/two-factor/blob/529c973deb4677e45cc243b1336dd718aa472611/providers/class.two-factor-fido-u2f.php#L70

`Notice: wp_register_script was called incorrectly. Scripts and styles should not be registered or enqueued until the wp_enqueue_scripts, admin_enqueue_scripts, or login_enqueue_scripts hooks. Please see Debugging in WordPress for more information. (This message was added in version 3.3.0.) in /srv/www/ona-journalists/public_html/wp-includes/functions.php on line 5167
`

### PR DESCRIPTION
**Pull request by kasparsd**
Fixes #361.

### CODE REVIEW

File: providers/class-two-factor-fido-u2f.php [hunk 1]
```diff
@@ -131,12 +120,27 @@ public function get_label() {
...
 
 	/**
-	 * Enqueue assets for login form.
+	 * Register script dependencies used during login and when
+	 * registering keys in the WP admin.
 	 *
-	 * @since 0.1-dev
+	 * @return void
 	 */
-	public function login_enqueue_assets() {
```
Comment (kasparsd) [inline providers/class-two-factor-fido-u2f.php:138]: This wasn't used anywhere.

File: providers/class-two-factor-fido-u2f.php [hunk 2]
```diff
@@ -131,12 +120,27 @@ public function get_label() {
...
+		wp_register_script(
+			'fido-u2f-api',
+			plugins_url( 'includes/Google/u2f-api.js', dirname( __FILE__ ) ),
+			null,
+			self::asset_version(),
+			true
+		);
+
+		wp_register_script(
+			'fido-u2f-login',
```
Comment (kasparsd) [inline providers/class-two-factor-fido-u2f.php:138]: This is used during our custom "login" page https://github.com/wordpress/two-factor/blob/a97c32b6639a3557fd149725b973615246549ffe/class-two-factor-core.php#L581-L726

File: providers/class-two-factor-fido-u2f.php [hunk 3]
```diff
@@ -131,12 +120,27 @@ public function get_label() {
...
+	 * registering keys in the WP admin.
 	 *
-	 * @since 0.1-dev
+	 * @return void
 	 */
-	public function login_enqueue_assets() {
-		wp_enqueue_script( 'fido-u2f-login' );
+	public static function enqueue_scripts() {
+		wp_register_script(
+			'fido-u2f-api',
```
Comment (kasparsd) [inline providers/class-two-factor-fido-u2f.php:130]: This is a dependency for `fido-u2f-admin` and `fido-u2f-login`.

File: providers/class-two-factor-fido-u2f.php [hunk 4]
```diff
@@ -73,21 +73,10 @@ protected function __construct() {
...
-
-		wp_register_script(
-			'fido-u2f-login',
-			plugins_url( 'js/fido-u2f-login.js', __FILE__ ),
-			array( 'jquery', 'fido-u2f-api' ),
-			self::asset_version(),
-			true
-		);
+		// Ensure the script dependencies have been registered before they're enqueued at a later priority.
+		add_action( 'admin_enqueue_scripts', array( __CLASS__, 'enqueue_scripts' ), 5 );
```
Comment (kasparsd) [inline providers/class-two-factor-fido-u2f.php:77]: Use action priority 5 to ensure the scripts have been registered before they're enqueued later at priority 10.

**Comment by rinatkhaziev** [comment]
This looks good, however,  is there a particular reason why we do script registration and enqueueing separately? I may be missing some context, but it seems to me that the complexity might be reduced a bit if combined into a single `wp_enqueue_script` call.

**Comment by kasparsd** [comment]
@rinatkhaziev That's a good question. I think the reason for that is that both admin and login screens rely on the same JS dependencies so we're able to reference those dependencies without having to enqueue them two multiple times and potentially mess up the file paths, versions, etc.

**Comment by rinatkhaziev** [comment]
Ah I see, thanks!

**Pull request merged**